File: Graphs/23/time_needed.py
class Solution:
    def numOfMinutes(self, n: int, headID: int, manager: List[int], informTime: List[int]) -> int:
        if n <= 1:
            return 0
        direct_sub = {}
        for i, ele in enumerate(manager):
            if ele > -1:
                if direct_sub.get(ele):
                    direct_sub[ele].append(i)
                else:
                    direct_sub[ele] = [i]

        timeFromHead = [0]*n

        for ele in direct_sub.get(headID):
            timeFromHead[ele] = informTime[headID]

        employees = direct_sub.get(headID)
        maxTime = informTime[headID]

        while len(employees) > 0:
            emp = employees.pop()
            informTimeToSub = timeFromHead[emp] + informTime[emp]
            maxTime = max(maxTime, informTimeToSub)
            if direct_sub.get(emp):
                for sub in direct_sub[emp]:
                    timeFromHead[sub] = informTimeToSub
                    employees.append(sub)
        return maxTime

s = Solution()
s.numOfMinutes(6, 2, [2, 2, -1, 2, 2, 2], [0, 0, 1, 0, 0, 0])

File: Graphs/23/author/time_needed.js
var numOfMinutes = function (n, headID, manager, informTime) {
  if (n <= 1) return 0;
  const subordinates = manager.map(() => []);
  for (let i = 0; i < manager.length; i++) {
    if (manager[i] > -1) {
      subordinates[manager[i]].push(i);
    }
  }

  return dfs(headID, subordinates, informTime);
};

var dfs = function (id, subordinates, informTime) {
  if (subordinates[id].length === 0) return 0;

  let max = 0;
  for (let i = 0; i < subordinates[id].length; i++) {
    max = Math.max(max, dfs(subordinates[id][i], subordinates, informTime));
  }
  return max + informTime[id];
};
